### PR TITLE
Tasks endpoints

### PR DESCRIPTION
**Pull request by hypesystem**
I focused a lot on the integration tests, taking a slightly different approach from what you have been doing. Hopefully you can see why :-) also I show how to use cucumber data tables, which I seem to recall you asked for an example of :-)

I also left some commends (TODOs) on the topics integration tests. Nothing of urgence, just for some point in the future. And up for discussion, of curse 😛 

The endpoints themselves are very much like the topic ones. The only innovation is the added ID-deconflicting I have made:

- IDs are unique, decided upon creation.
- IDs are built from the title of a task. "Awesome Title" -> "awesome-title"
- Titles of tasks can be changed afterwards, but the ID stays the same.
- When a new card is created that would clash with an existing ID, it simply appends "-1" (and then "-2", "-3", etc) to the ID.

This code could probably be extracted and reused in topics :smile:

### CODE REVIEW

**Comment by hypesystem** [comment]
README has not been updated yet. Depending on what you say to the rest  of the code, I will update it.

Also no vbump yet.

File: features/step_definitions/tasks.js [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+    Given(/^the circle has a task "([^"]+)"(, which I own)?, with the following data:$/, function(taskName, isOwner, table, callback) {
+        if(isOwner) {
+            let data = table.rowsHash();
+            data.title = taskName;
+            return ensureTask.call(this, data, callback);
+        }
+        else {
+            callback(null, 'pending');
+        }
```
Comment (ath88) [inline features/step_definitions/tasks.js]: http://blog.hypesystem.dk/ifelse-ifreturn

File: features/step_definitions/tasks.js [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+            callback();
+        });
+    });
+
+    Then(/^no task with ID "([^"]+)" should exist$/, function(taskId, callback) {
+        requestTaskList.call(this, error => {
+            if(error) {
+                return callback(error);
+            }
+            let matchingTask = this.body.tasks.find(task => task.taskId == taskId);
```
Comment (ath88) [inline features/step_definitions/tasks.js]: please put parens around arguments in arrow functions :heart:

File: lib/circles/tasks/create.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                let existingCanonicalIds = result.rows.map(row => row.id);
+                let newTaskId = determineNewTaskId(existingCanonicalIds, req.circle.circleId, task.taskId);
+                task.taskId = newTaskId;
+                task.canonicalTaskId = `${req.circle.circleId}/${newTaskId}`;
+                insertTask(req, task, callback);
+            });
+        }
+        callback(error, result);
+    });
+}
```
Comment (ath88) [inline lib/circles/tasks/create.js:51]: This still leaves a problem with race conditions and can return an error of "tasks_pkey"-type. Could be fixed, not important now though.

File: lib/circles/topics/update.js [hunk 4]
```diff
@@ -5,7 +5,7 @@ module.exports = function(req, res) {
 
     const topicParams = {
         title: req.body.title,
-        owner: req.body.userId,
+        owner: req.body.userId, //TODO: Shouldn't this get req.body.owner? (Seems odd to call it userId in the request!)
```
Comment (ath88) [inline lib/circles/topics/update.js:8]: Yes. I agree.

File: lib/circles/topics/update.js [hunk 5]
```diff
@@ -21,7 +21,7 @@ module.exports = function(req, res) {
         return res.status(400).send({ status: "There are no values in request" });
     }
 
-    if(cleanTopicParams.stage && !stageEnum.includes(topicParams.stage)) {
+    if(cleanTopicParams.stage && !stageEnum.includes(topicParams.stage)) { // TODO: Shouldn't topicParams here be cleanTopicParams?
```
Comment (ath88) [inline lib/circles/topics/update.js:24]: Yup.

File: lib/circles/topics/update.js [hunk 6]
```diff
@@ -36,6 +36,7 @@ module.exports = function(req, res) {
         const updatedTopic = _.merge(result.rows[0].data, cleanTopicParams);
 
         if(cleanTopicParams.title && cleanTopicParams.title !== updatedTopic.title) return res.status(501).send({ status: "Renaming topics is not supported" });
+        //TODO: Is there any reason not to support renaming anymore? As long as the ID stays intact :-)
```
Comment (ath88) [inline lib/circles/topics/update.js:39]: You are correct - feel free to implement. :)

**Review by ath88 — APPROVED** [review]
Look good - few comments. :)

File: features/step_definitions/tasks.js (hunk 1 above)
Comment (hypesystem) [reply features/step_definitions/tasks.js]: flab

**Comment by hypesystem** [comment]
Fixed issues related to tasks.

Left TODOs in topics for you my friend @ath88

**Pull request merged**